app/scripts/migrations/127.ts

```diff
@@ -0,0 +1,414 @@
+        return endpoints;
+      },
+      [],
+    );
+
+    // If there were no valid unique endpoints, then omit the network
+    // configuration for this chain id. The network controller requires
+    // configurations to have at least 1 endpoint.
+    if (rpcEndpoints.length === 0) {
+      return acc;
+    }
+
+    // Use the tie breaker network as the default rpc endpoint
+    const defaultRpcEndpointIndex = Math.max(
+      rpcEndpoints.findIndex(
+        (endpoint) => endpoint.networkClientId === tieBreaker?.id,
+      ),
+      // Or arbitrarily default to the first endpoint if we don't have a tie breaker
+      0,
+    );
+
+    // Calculate the unique array of non-empty block explorer urls
+    const blockExplorerUrls = [
+      ...networks.reduce((urls, network) => {
+        if (
+          isObject(network.rpcPrefs) &&
+          typeof network.rpcPrefs.blockExplorerUrl === 'string' &&
+          network.rpcPrefs.blockExplorerUrl
+        ) {
+          urls.add(network.rpcPrefs.blockExplorerUrl);
+        }
+        return urls;
+      }, new Set()),
+    ];
+
+    // Use the tie breaker network as the default block explorer, if it has one
+    const defaultBlockExplorerUrlIndex =
+      blockExplorerUrls.length === 0
+        ? undefined
+        : Math.max(
+            blockExplorerUrls.findIndex(
+              (url) =>
+                isObject(tieBreaker?.rpcPrefs) &&
+                url === tieBreaker.rpcPrefs.blockExplorerUrl,
+            ),
+            // Or arbitrarily default to the first url
+            0,
+          );
+
+    // Use the cononical network name and currency, if we have constants for them.
+    // Otherwise prefer the tie breaker's name + currency, if it defines them.
+    // Otherwise fall back to the name + currency from arbitrary networks that define them.
+    const name =
+      NETWORK_TO_NAME_MAP[chainId as keyof typeof NETWORK_TO_NAME_MAP] ??
+      tieBreaker?.nickname ??
+      networks.find((n) => n.nickname)?.nickname;
+
+    const nativeCurrency =
+      CHAIN_ID_TO_CURRENCY_SYMBOL_MAP[
+        chainId as keyof typeof CHAIN_ID_TO_CURRENCY_SYMBOL_MAP
+      ] ??
+      tieBreaker?.ticker ??
+      networks.find((n) => n.ticker)?.ticker;
+
+    acc[chainId] = {
+      chainId,
+      rpcEndpoints,
+      defaultRpcEndpointIndex,
+      blockExplorerUrls,
+      ...(defaultBlockExplorerUrlIndex !== undefined && {
+        defaultBlockExplorerUrlIndex,
+      }),
+      name,
+      nativeCurrency,
+    };
+    return acc;
+  }, {});
+
+  // Ensure that selectedNetworkClientId points to
+  // some endpoint of some network configuration.
+  let selectedNetworkClientId = Object.values(networkConfigurationsByChainId)
+    .flatMap((n) =>
+      isObject(n) && Array.isArray(n.rpcEndpoints) ? n.rpcEndpoints : [],
+    )
+    .find(
+      (e) => e.networkClientId === networkState.selectedNetworkClientId,
+    )?.networkClientId;
+
+  // It may not, if its endpoint was not well formed
```

Comment:
Invalid and duplicate endpoints are an edge case since the UI already validates both.  But playing it safe since the network controller introduced `uri-js` for validity + equality which could be stricter e.g. it considers different casing to be equal.

If the duplicate was within the same chain id, we do prefer the one that was selected due to this guy:

```
// If there's a duplicate RPC url within a chain, and one of the
// networks is the globally selected network, prefer to use its network
// client id so that `selectedNetworkClientId` can remain unchanged.
```

The case bring commented on is for when the selected endpoint had to be removed due to invalidity.  Or it was a duplicate across chain ids, where we just take the first one we see.  You could try to do a similar preference across chain ids.  But then you can't just process one network at a time, since you'd have to go back and re-process the old one whose endpoint you kicked out.  So trying to avoid complexity for an edge case of an edge case.

Reply:
I improved the handling of duplicates across chain ids in https://github.com/MetaMask/metamask-extension/pull/26433/commits/74f01a334588c854da34a42f52a7ea8067bbeadd

Now if there's a duplicate, whether within or across chains, it will prefer to use the one that is the selected network.  If none are the selected network, it takes the first duplicate it encounters

Reply:
Note: Invalid + duplicate endpoints are an edge case since the UI validates both already.  But (1) I don't know if that validation was already there, (2) the network controller won't initialize if validation fails, and (3) it introduced `uri-js` for validity + equality which can be stricter e.g. it considers different casing to be equal.